### PR TITLE
DrivAerML: higher LR + linear warmup (widening the stable LR basin)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

LR=5e-4 was found to be a sharp optimum on DrivAerML — small deviations cause catastrophic divergence. However, this was tested without warmup. Linear warmup gradually builds LR from 0, which can stabilize higher peak LRs that would otherwise diverge from cold start.

A higher peak LR (7e-4 or 1e-3) with 20-epoch linear warmup to the peak, then cosine decay with T_max=30, could reach a deeper loss basin faster. The warmup allows the optimizer to find a stable trajectory before hitting the high LR region.

PRs #3069 (5-epoch warmup) and #3070 (10-epoch warmup) test warmup at the champion lr=5e-4. This PR tests warmup at HIGHER LR to see if it widens the stable operating range.

## Instructions

Implement linear LR warmup in train.py using a LinearLR scheduler chained before the cosine annealing. Run two configurations:

**Run 1: lr=7e-4 with 20-epoch warmup**
- Start LR: 1e-5, ramp linearly to 7e-4 over 20 epochs
- Then: CosineAnnealingLR with T_max=30 and eta_min=1e-6
- Use `torch.optim.lr_scheduler.SequentialLR([LinearLR(..., start_factor=1e-5/7e-4, end_factor=1.0, total_iters=20), CosineAnnealingLR(T_max=30)], milestones=[20])`

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 7e-4 \
  --cosine-t-max 30 \
  --warmup-epochs 20 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb_group dm-higher-lr-warmup
```

**Run 2: lr=1e-3 with 20-epoch warmup**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 1e-3 \
  --cosine-t-max 30 \
  --warmup-epochs 20 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb_group dm-higher-lr-warmup
```

If train.py doesn't have a --warmup-epochs flag, add it: implement SequentialLR with LinearLR for warmup_epochs then cosine for remainder. No --grad-clip, no --weight-decay.

Report best val_primary/surface_rel_l2_pct for both runs.

## Baseline

- **val_primary/surface_rel_l2_pct:** 3.997% at epoch 467
- **Config:** 4L/512d/8H, AdamW lr=5e-4, T_max=30, no-EMA, Fourier, no gc, no WD, no warmup
- **W&B run:** bht6h42t
- **Reproduce:** `cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT historia:

## Results

### DrivAerML: higher LR + 20-epoch linear warmup

| Metric | Baseline (lr=5e-4, no warmup) | Run 1: lr=7e-4 | Run 2: lr=1e-3 |
|--------|-------------------------------|-----------------|-----------------|
| **val best `surface_rel_l2_pct`** | **3.997%** @ ep467 | **19.82%** @ ep15 | **17.75%** @ ep32 |
| Warmup divergence epoch | N/A | ep18 (LR ~6.3e-4) | ep17 (LR ~8.5e-4) |
| Cosine phase stability | Stable for 467 epochs | Oscillating 28-56% | Terminal divergence ep46+ |
| Final state | Converged | Oscillating | Grad norms >20,000 |

**Run 1 — lr=7e-4 with 20-epoch warmup:**
- W&B run: `8swgnhho` (historia/dm-lr7e4-warmup20, group: dm-higher-lr-warmup)
- Best val: **19.82%** at epoch 15 (during warmup, LR ~5.3e-4)
- Warmup phase smooth until epoch 18 (LR ~6.3e-4) — metric spiked from 20.9% to 75.7%
- Cosine phase: volatile oscillation 28-56%, never recovered to warmup-phase bests
- Grad norms elevated (3.5-7.8) throughout cosine phase
- Peak memory: ~6207 MiB (GPU 0)

**Run 2 — lr=1e-3 with 20-epoch warmup:**
- W&B run: `87ecgqp8` (historia/dm-lr1e3-warmup20, group: dm-higher-lr-warmup)
- Best val: **17.75%** at epoch 32 (during cosine phase)
- Warmup phase smooth until epoch 17 (LR ~8.5e-4) — metric spiked from 19.9% to 108.6%
- Cosine phase initially recovered (17.7% at ep32), then catastrophic divergence from epoch 46
- Terminal state: grad norms 20,000-26,000+, metric stuck at 63-73%, irrecoverable
- Peak memory: ~6207 MiB (GPU 1)

**Commands used:**
```bash
# Run 1
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 7e-4 --cosine-t-max 30 --warmup-epochs 20 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group dm-higher-lr-warmup --agent historia --wandb-name "historia/dm-lr7e4-warmup20"

# Run 2
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 1e-3 --cosine-t-max 30 --warmup-epochs 20 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group dm-higher-lr-warmup --agent historia --wandb-name "historia/dm-lr1e3-warmup20"
```

### Implementation

Added `--warmup-epochs` flag to train.py. When set >0, uses `torch.optim.lr_scheduler.SequentialLR` to chain:
1. `LinearLR` (start_factor=1e-5/lr, end_factor=1.0, total_iters=warmup_epochs*steps_per_epoch)
2. `CosineAnnealingLR` (T_max=config.cosine_t_max)

### What happened

**HYPOTHESIS FALSIFIED.** Linear warmup does NOT widen the stable LR basin on DrivAerML.

Both runs show the same pattern:
1. **Warmup phase (ep1-17/18):** Smooth, promising descent. Both models trained well at intermediate LRs (4e-4 to 6e-4 range). Best warmup-phase metrics were 19.8% (Run 1) and 19.9% (Run 2).
2. **Warmup instability (ep17-20):** Once LR exceeded ~6e-4, both models diverged catastrophically. Run 1 spiked to 75.7%, Run 2 spiked to 108.6%. The instability threshold at ~6e-4 is the same whether warmup is used or not.
3. **Cosine phase:** The cosine schedule cycles LR repeatedly through the unstable region (>6e-4). Run 1 survived but oscillated wildly (28-56%). Run 2 initially recovered to 17.75% but then terminally diverged with grad norms >20,000.

**Key insight:** The DrivAerML LR stability boundary at ~6e-4 is a property of the loss landscape, not the optimization trajectory. Warmup doesn't change the loss landscape — it only changes how the optimizer approaches it. Once the cosine schedule pushes LR above the stability threshold, the same divergence occurs regardless of warmup. The champion's lr=5e-4 stays below this boundary for the entire training run, which is why it's stable for 467+ epochs.

**Interesting observation:** During warmup, both models were performing identically to the champion at equivalent LR levels. The problem isn't reaching the target LR — it's staying at it.

### Verdict

**NEGATIVE.** Neither run beats the baseline (19.82% / 17.75% vs 3.997%). Warmup is useful infrastructure (kept in train.py) but does not enable higher peak LRs on DrivAerML.

### Suggested follow-ups

1. **Warmup + grad-clip + higher LR:** The instability is a gradient explosion. Combining warmup with grad-clip=1.0 at lr=7e-4 or lr=1e-3 could stabilize the high-LR regime.
2. **Warmup + lower peak LR (e.g. lr=6e-4):** Warmup at 6e-4 (just below the stability boundary) might allow the model to find slightly deeper basins than cold-start at 5e-4.
3. **Cosine with eta_min > 0:** Instead of cycling LR to 0, set eta_min to prevent the model from ever reaching the unstable high-LR region during cosine restarts. This is different from just lowering the peak LR.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3119 because ADVISOR: Both LR variants catastrophically fail — lr=7e-4 reaches only 19.82% (oscillates 28-56%), lr=1e-3 gets 17.75% then diverges to NaN at ep46. This is 5-6x worse than the 3.833% baseline. Your analysis is correct: the DM stability boundary is ~6e-4 regardless of warmup. Combined with prior results (4e-4, 4.5e-4, 5.5e-4, 7e-4, 1e-3 all worse), the LR direction is fully exhausted for DrivAerML. lr=5e-4 is confirmed as a sharp optimum..

**Comment by tcapelle** [comment]
STUDENT historia:

## Test metrics update

Both training runs completed. Final best-checkpoint evaluation:

| Run | Best Epoch | Val `surface_rel_l2_pct` | Test `surface_rel_l2_pct` |
|-----|-----------|--------------------------|---------------------------|
| **Run 1 (lr=7e-4)** | 15 | 19.82% | 19.61% |
| **Run 2 (lr=1e-3)** | 37 | 16.56% | 17.23% |
| **Baseline** | 467 | 3.997% | — |

Note: Run 2's best val was actually 16.56% at epoch 37 (slightly better than the 17.75% I tracked during monitoring). The test metric (17.23%) confirms good generalization at the best checkpoint. Both runs are still far from the 3.997% baseline.
